solutions/usaco-380.mdx

```diff
@@ -0,0 +1,132 @@
+
+## Implementation
+
+<LanguageSection>
+
+<JavaSection>
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class ccski {
+	static int n, m;
+	static int startI, startJ; // Stores starting position of each floodfill
+	static int[][] course; // Stores the skiing course heights
+	static boolean[][] waypoints; // Stored as booleans instead of 1s and 0s
+	static boolean[][] vis; // Visited array for floodfill
+
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio("ccski");
+
+		n = io.nextInt();
+		m = io.nextInt();
+
+		int minHeight = Integer.MAX_VALUE;
+		int maxHeight = Integer.MIN_VALUE;
+
+		course = new int[n][m];
+		for (int i = 0; i < n; i++) {
+			for (int j = 0; j < m; j++) {
+				course[i][j] = io.nextInt();
+				minHeight = Math.min(minHeight, course[i][j]);
+				maxHeight = Math.max(maxHeight, course[i][j]);
+			}
+		}
+		
+		waypoints = new boolean[n][m];
+		for (int i = 0; i < n; i++) {
+			for (int j = 0; j < m; j++) {
+				if (io.nextInt() == 1) {
+					waypoints[i][j] = true;
+
+					// Keep one of the waypoints as the starting position
+					startI = i;
+					startJ = j;
+				} else {
+					waypoints[i][j] = false;
+				}
+			}
+		}
+
+		// Binary search for the minimum possible value of d that works
+		// We can set "hi" to maxHeight - minHeight because that will be the
+		// maximum value of d we need to try
+		int lo = 0;
+		int hi = maxHeight - minHeight;
+		int minD = -1;
+		while (lo <= hi) {
+			int d = (lo + hi) / 2;
+			if (reachable(d)) {
+				minD = d;
+				hi = d - 1;
+			} else {
+				lo = d + 1;
+			}
+		}
+		
+		io.println(minD);
+		io.close();
+	}
+
+	// i and j store current position
+	// d stores current value of d
+	// prevHeight stores the height of the previous cell
+	static void floodfill(int i, int j, int d, int prevHeight) {
+		// Check if we are out of bounds
+		if (i < 0 || i >= n || j < 0 || j >= m) {
+			return;
+		}
+		// Check if current position has already been visited
+		if (vis[i][j]) {
+			return;
+		}
+		// Check if current position can be visited from previous position
+		if (Math.abs(course[i][j] - prevHeight) > d) {
+			return;
+		}
+
+		// Mark position as visited if all checks pass
+		vis[i][j] = true;
+
+		// Visit each adjacent cell
+		floodfill(i + 1, j, d, course[i][j]);
+		floodfill(i - 1, j, d, course[i][j]);
+		floodfill(i, j + 1, d, course[i][j]);
+		floodfill(i, j - 1, d, course[i][j]);
+	}
+
+	static boolean reachable(int d) {
+		// Reset visited array and begin floodfill (DFS) from start position
+		vis = new boolean[n][m];
+		floodfill(startI, startJ, d, course[startI][startJ]);
+		
+		// Check each cell: if it is a waypoint and it hasn't been visited,
+		// we know not all waypoints are reachable from one another
+		for (int i = 0; i < n; i++) {
+			for (int j = 0; j < m; j++) {
+				if (waypoints[i][j] && !vis[i][j]) {
+					return false;
+				}
+			}
+		}
+		
+		return true;
+	}
+
+	CodeSnip{Kattio}
```

Comment:
add // in front to make it actually work

Reply:
This is not necessary, but it's fine if you want to add it